### DISCUSSION TITLE
Pluck method with fixed key

### DISCUSSION THREAD

**Discussion by migueliriano** [laravel/framework#48953] [Ideas]
What do you think about extending [Pluck](https://laravel.com/docs/10.x/collections#method-pluck) to retrieve all of the values with a fixed key ?

For example in this collection, I want the `name` values but with a fixed key called `product_name`

Example code:

```php

$collection = collect([
    ['product_id' => 'prod-100', 'name' => 'Desk'],
    ['product_id' => 'prod-200', 'name' => 'Chair'],
]);

// Current Solution:
$plucked = $collection->map(fn($collect) => ['product_id' => $collect->name]);

// Proposal solution extending `Pluck` :
$plucked = $collection->pluck(['product_name', 'name']);

// Expected output
// ['product_name' => 'Desk', 'product_name'  => 'Chair']
```

So the idea as you can see in the example, is extending the Pluck method of the collection to accept an array where the first index is the fixed key value `product_name` and the other is the column that you want to retrieve `name`.

**Comment by shaedrich** [comment]
Unfortunately, passing an array to the `pluck` method is already possible. The method then returns each column specified in the passed array. So your solution wouldn't work without introducing a breaking change and replacing existing behavior.

**Reply by migueliriano** [reply]
Thank you for this information, for some reason `pluck` is not working as you mentioned,  when I pass an array it does not return each column, actually is returns `null`

Following the same example data:

```php
$collection = collect([
		['product_id' => 'prod-100', 'name' => 'Desk'],
		['product_id' => 'prod-200', 'name' => 'Chair'],
]);
$collection->pluck(['product_id', 'name'])->toArray();

//  output
// array(2) { [0]=>  NULL  [1]=>  NULL}
```

Also, I don't see that usage in the [Laravel 10 doc for `pluck`](https://laravel.com/docs/10.x/collections#method-pluck), maybe is a deprecated feature 🤔

**Reply by migueliriano** [reply]
Also, thinking more about this, we already have the second parameter called `key`
```php
   public function pluck($value, $key = null)
```

Maybe we can extend that parameter to do one of the next options:

1. Use `$key` parameter as a static  key if the column doesn't exist
2. If doing the first option can be risky because we are changing the behavior of that parameter, maybe we can add a third boolean parameter like `fixedKey` that is `false`  by default, and if it's `true` that  going to use the key as the fixed value instead of search another column

```php
   public function pluck($value, $key = null, $fixedKey = false)
```

Examples of each point: 

```php

$collection = collect([
		['product_id' => 'prod-100', 'name' => 'Desk'],
		['product_id' => 'prod-200', 'name' => 'Chair'],
]);

// Example option 1: `product_name` doesn't exist in the collection, so going to be automatically added as fixed key
$collection->pluck('name', 'product_name')->toArray();

//Example option 2: The third `$fixedKey` is set to `true`, the second parameter will be used as a fixed key for the array result
$collection->pluck('name', 'product_name', true)->toArray();

// Same output for both points
// ['product_name' => 'Desk', 'product_name'  => 'Chair']

```

**Reply by shaedrich** [reply]
3. [BREAKING CHANGE] Make `$key`'s data type a `callable` by default if given:
```php
$collection->pluck('name', fn($entry) => $entry->name)->toArray(); // => Returns 'Desk' and 'Chair' as keys respectively
$collection->pluck('name', 'product_name')->toArray(); // => Returns 'product_name' as key for each entry
```
